[WIP] Add serializer interface

matthiasbeyer opened the pull request:
Targets #161.

matthiasbeyer committed:
Add header for ws_transaction_command type

matthiasbeyer committed:
Add header for ws_transaction type

neithernut commented on src/command/command_argument.h:
```diff
@@ -0,0 +1,43 @@
+ */
+struct ws_transaction_command_argument {
+    struct ws_object obj;
+    struct ws_value* argv;
```
We have a stack/dynamic range of registers. You don't want new values but the addresses of existing registers.
Use an array of `size_t` or some ints (add the length or come up with some termination symbol)

neithernut commented on src/command/transaction.h:
```diff
@@ -0,0 +1,43 @@
+ * Transaction type, extends the array type
+ */
+struct ws_transaction {
+    struct ws_object ary;
```
You meant to have a `struct ws_array` here, didn't you?

matthiasbeyer replied:
Jep.

matthiasbeyer committed:
Add serializer type interface

neithernut commented on src/command/command.h:
```diff
@@ -0,0 +1,50 @@
+    char* callstr;
+
+    struct {
+        size_t ary;
```
How is this supposed to hold multiplel params' positions?

matthiasbeyer replied:
Argh.

matthiasbeyer committed:
Add stubs

matthiasbeyer commented:
@neithernut functions missing?

neithernut commented on src/command/command.h:
```diff
@@ -0,0 +1,50 @@
+ * Transaction command
+ */
+struct ws_transaction_command {
+    struct ws_object obj;
```
A single command does not need to be an object.
If it's only for the sake of storing it in an array, you might use a `wl_array` instead for the transaction.

matthiasbeyer replied:
The transaction is already an array. So I guess we can just remove the inheritance and then we're fine here?

neithernut replied:
Yep. But we discussed whether `ws_array`s hold `ws_objects` or other data, too, if you remember.

neithernut commented on src/serialize/module.h:
```diff
@@ -28,6 +28,42 @@
+ * Serializer type
+ */
+struct ws_serializer {
+    struct ws_object obj; //!< Supertype
```
A serializer isn't neccessarily an object.
We have exactely one serializer per command processor. We may have different types of deserializers at some point, but even in that case having a struct with callback fits better than making it an object.

matthiasbeyer replied:
We discussed this yesterday and yuo said it is an object!

neithernut replied:
I think I said that it is not derived from `ws_object`, but that it has an internal state, which serves as a context for yaijl.
object != object.

neithernut commented on src/command/command.h:
```diff
@@ -0,0 +1,50 @@
+struct ws_transaction_command {
+    struct ws_object obj;
+
+    char* callstr;
```
It would be better to resolve this in the serializer.
Add a new function typedef like

``` C
int (*command)(struct ws_arguments const*)
```

Of course, to do this you have to make the struct below unanonymous.

matthiasbeyer replied:
For what exactly do you wnat me to add a callback here?

neithernut commented:
I'd like to have this merged once the abstract definition and some basic infrastructure are done. This way, we can do this in paralell.

matthiasbeyer commented on src/command/command.h:
```diff
@@ -0,0 +1,50 @@
+    char* callstr;
+
+    struct {
+        size_t* ary;
```
Shouldn't we use fixed-size elements here? I would like to suggest `uint32_t`.

matthiasbeyer committed:
Add interface definition for serialize backend

matthiasbeyer committed:
Add stubs for serialize backend

matthiasbeyer committed:
Add interface definition for deserialize backend

matthiasbeyer committed:
Add stubs for deserialize backend

matthiasbeyer committed:
Add (de)serializer backend members for serializer type

matthiasbeyer committed:
Add abstract backend type for serializer backends

matthiasbeyer committed:
Fixup: Abstract backend interface functionality

matthiasbeyer commented:
I added some more infrastructure and rewrote almost the complete history. The inheritance was split up once again and now we have:
- A serializer type
- A serializer backend type (abstract)
- A serializer serialize backend type (abstract)
- A serializer deserialize backend type (abstract)

The (json-, xml-, foobar-) backend (both serialize and deserialize) types are supposed to inherit from the latter two.

I'm not completely sure whether the last point is sane, as serializing and de-serializing could happen on the same backend object, _but_ what I had in mind with this is, that serializing and deserializing is a completely different task and with this setup, we could have different instances of the serializer doing serializing and deserializing _in parallel_, so a deserializing-job does not necessarily block a serializing job and the other way round. So this setup supports the threading and therefor the speed of the application.

---

More patches are coming.

matthiasbeyer commented on src/command/command.h:
```diff
@@ -0,0 +1,48 @@
+    char* callstr;
+
+    struct {
+        size_t* ary;
```
I'll re-add this comment here: Shouldn't we use a fixed-size type, like `uint32_t`? Plugins would be much more portable this way!

neithernut replied:
The pointer alone means you are bound the the concrete architecture concerning binary representation. `size_t` is defined for one architecture, so it's sane.

neithernut commented:
Introducing the possibility to perform serialisation and deserialisation in different threads is a good idea. However, I don't see any point in having them both based on a general serializer backend.

neithernut commented on src/command/transaction.h:
```diff
@@ -0,0 +1,43 @@
+ * Transaction type, extends the array type
+ */
+struct ws_transaction {
+    struct ws_array ary;
```
As described earlier: making this a `wl_array` makes more sense.

matthiasbeyer replied:
It _is_ a `wl_array`!

neithernut replied:
It sais `ws_array`, which implied it holds objects.

matthiasbeyer replied:
Narg, I misread, I'm sorry. ("wl" vs. "ws", you know?)

neithernut commented on src/serialize/backend.h:
```diff
@@ -0,0 +1,112 @@
+ * This is the supertype of all backend types for the serializer
+ */
+struct ws_serializer_backend {
+    struct ws_object obj; //!< @protected Base class.
```
As described before: neither serializers nor a deserializers are _this kind of object_.

matthiasbeyer replied:
The advantages for having a `ws_object` here are clear: You can do locking without having to implement it again as well as you have a runnable object here, so you can simply call `ws_object_run(serializer)` here. The advantages outperform the disadvantages (which are only overhead, or is there something else) here.

neithernut replied:
Simply nope.

First thing is you introduce a transaction getter _and_ a transaction setter, when you really need only one of them at the same time. Second: we defined the interface of the serializer to do the following:
- You lik it to a connection buffer
- In the libev handler callback, you try get a _message_. If a _message_ could be extracted, it returns that message. If data is missing, it returns NULL.

This makes callback design much more simple. Eg. you don't have any need for message buffers, queues or any other stuff like that.

Secondly: the code duplication is minimal, as the serializer and the deserializer each have a callback for exactly one task, and, maybe, a pointer to the `ws_connbuf` it's associated with.
That's minimal code-duplication.

matthiasbeyer replied:
The getter _and_ setters were added because of the idea of doing `ws_object_run()` on the object. The dereived classes do something, either transforming a transaction to a buffer or the other way round. So If you put a buffer into the backend, you can run it and get a transaction from it as well as the other way round.

That was my idea when writing this.

If I understand things right, you want me to implement an extended object type table which holds the callbacks for what exactly? You need _either_ to pass a buffer _or_ a transaction and get the other one out of the backend. How would you implement an interface for this?

(your patches to this are welcome)

neithernut replied:
At this time, I'll think I'll open my own PR.
That is, _after_ I've done the things important for reaching the _current_ milestone.

neithernut commented on src/serialize/backend.h:
```diff
@@ -0,0 +1,112 @@
+    struct {
+        void* b;
+        size_t bsize;
+    } buffer; //!< @protected Backend-Related buffer
```
Having an additional buffer here doesn't make any sense. We can work on the corresponding `ws_connbuf` in the connector directly.

matthiasbeyer replied:
True.

matthiasbeyer commented:
> Introducing the possibility to perform serialisation and deserialisation in different threads is a good idea. > However, I don't see any point in having them both based on a general serializer backend.

The point I made this is because they share the same variables. They both operate on a transaction and a buffer, just in different manner. So it is much more convenient to implement a supertype for them and providing the functionality for setting/getting members than implementing everything two times.

neithernut commented on src/serialize/backend.h:
```diff
@@ -0,0 +1,112 @@
+        void* b;
+        size_t bsize;
+    } buffer; //!< @protected Backend-Related buffer
+};
```
How is this supposed to work, anyways? You didn't even introduce callbacks.

matthiasbeyer replied:
Not by now. The backend type wouldn't introduce callbacks, though. The callbacks are only set in the actual implementation, which are not covered yet at all. For example in `struct ws_serializer_serialize_backend_json` or something.

The `struct ws_serializer_backend` is a abstract/virtual type.

matthiasbeyer committed:
Fixup: transaction type should be a wl_array

matthiasbeyer committed:
Fixup: Use ws_connbuf type for buffer in abstract backend(.h)

matthiasbeyer commented:
Somehow, the latest fix compiles. But `struct ws_connbuf` is not defined in the source, so why the hell does it compile?

neithernut commented:
If you define a pointer to a struct, the struct is implicitely forward declarated.
And as soon as you have a pointer, you don't need the size of the struct for compilation.

matthiasbeyer commented:
Thanks for the explanation. But shouldn't that fail when linking?

neithernut commented:
Only if you deref the struct.

matthiasbeyer commented:
:angry:
